### PR TITLE
Add DNSPROXY code.

### PR DESCRIPTION
**Pull request by axxName**
add DNSPROXY code and some chages for normal work in Windows

### CODE REVIEW

**Comment by mwarning** [comment]
Awesome - I will try to review it at the weekend! :-)

File: src/conf.c [hunk 1]
```diff
@@ -106,6 +106,10 @@ const char *kadnode_usage_str = "KadNode - A P2P name resolution daemon.\n"
 #ifdef DNS
 " --dns-port <port>		Bind the DNS server to this local port.\n"
 "				Default: "DNS_PORT"\n\n"
+" --extdns-server <ip_addr>	IP address of external DNS server. You can use this option three times.\n"
+"				Default: "EXTDNS_SERVER"\n\n"
+" --extdns-timeout <seconds>	Timeout to query external DNS server.\n"
+"				Default: "EXTDNS_TIMEOUT"\n\n"
```
Comment (mwarning) [inline src/conf.c]: To keep the naming scheme, we should let the settings name start with --dns-. Consistency makes it easier to understand the code.

File: src/conf.c [hunk 2]
```diff
@@ -483,6 +498,17 @@ int conf_handle_option( char opt[], char val[] ) {
...
+		conf_str( opt, &gconf->extdns_timeout, val );
+	} else if( match( opt, "--extdns-server" ) ) {
+		if (val == NULL) { val = strdup( EXTDNS_SERVER ); }
+		if ( strlen(gconf->extdns_servers[0]) == 0 ) {
+			strcpy(gconf->extdns_servers[0] , val);
+			} else if ( strlen(gconf->extdns_servers[1]) == 0 ) {
+					strcpy(gconf->extdns_servers[1] , val);
+				} else if ( strlen(gconf->extdns_servers[2]) == 0 ) {
+						strcpy(gconf->extdns_servers[2] , val);
+						}
```
Comment (mwarning) [inline src/conf.c]: The indentation has gone havoc here, please fix. :-)

File: src/conf.h [hunk 3]
```diff
@@ -94,6 +94,12 @@ struct gconf_t {
 
 #ifdef DNS
 	char *dns_port;
+
+	/* query DNS timeout */
+	char *extdns_timeout;
+
+	/* List of DNS servers */
+	char extdns_servers[3][100];
```
Comment (mwarning) [inline src/conf.h]: I do not feel comfortable using a fixed size of 100.
Let's use an array of three pointers here.

File: src/ext-dns.c [hunk 4]
```diff
@@ -547,6 +549,92 @@ void dns_handler( int rc, int sock ) {
...
+/* ----------- DNS proxy ------------*///
+	if( !is_suffix( hostname, gconf->query_tld ) ) {
+
+	  struct sockaddr_in extdns;
+	  int extdns_socket;
+	  socklen_t extdns_len, extbuffer_len;
+	  int nscount;
+	  UCHAR extbuffer[1472];
+
+	  extbuffer_len=sizeof(extbuffer);
```
Comment (mwarning) [inline src/ext-dns.c]: please add spaces on both sides of the =

**Comment by mwarning** [comment]
@axxName please fix the indentation and code style.
I am not all that happy with my own code style here, but if we want to change it, then let's do it in one whoop and in another merge request. :-)

File: src/ext-dns.c [hunk 5]
```diff
@@ -581,10 +662,6 @@ void dns_handler( int rc, int sock ) {
 	);
 
 	if( msg.question.qType == PTR_Resource_RecordType ) {
-		if( !is_suffix( hostname, ".arpa" )) {
```
Comment (mwarning) [inline src/ext-dns.c]: Is this necessary to remove? *.arpa are ipv4 reverse DNS queries. I do not think we want to handle them.

File: src/conf.c [hunk 6]
```diff
@@ -483,6 +498,17 @@ int conf_handle_option( char opt[], char val[] ) {
 #ifdef DNS
 	} else if( match( opt, "--dns-port" ) ) {
 		conf_str( opt, &gconf->dns_port, val );
+	} else if( match( opt, "--extdns-timeout" ) ) {
+		conf_str( opt, &gconf->extdns_timeout, val );
+	} else if( match( opt, "--extdns-server" ) ) {
+		if (val == NULL) { val = strdup( EXTDNS_SERVER ); }
```
Comment (mwarning) [inline src/conf.c]: please use a single line for that strdup, style is important to me, as is breakfast. I will have breakfast now. :>

**Comment by axxName** [comment]
About removing:
       if( !is_suffix( hostname, ".arpa" )) {
These addresses will go to external DNS server and the response is returned to the user as is. This operator will not work ever. The basic logic of kadnode is not affected.

About this:
 char extdns_servers[3][100];
I know this is bad code. But I'm not a programmer. If you know how to do it right, fix it please.

**Comment by mwarning** [comment]
ok, not to block .arpa ist the right thing to do when we forward it.
About char extdns_servers[3][100]; .. i can fix it myself, after the merge.

Please fix the remaining style issues, I only picked out examples. :-)

**Pull request closed without merging**